python_modules/dagster/dagster/_core/definitions/metadata/source_code.py

```diff
@@ -0,0 +1,265 @@
+import inspect
+import os
+from pathlib import Path
+from typing import (
+    TYPE_CHECKING,
+    Any,
+    Callable,
+    List,
+    Optional,
+    Sequence,
+    Union,
+)
+
+import pydantic
+
+import dagster._check as check
+from dagster._annotations import experimental
+from dagster._model import DagsterModel
+from dagster._serdes import whitelist_for_serdes
+
+from .metadata_set import (
+    NamespacedMetadataSet as NamespacedMetadataSet,
+    TableMetadataSet as TableMetadataSet,
+)
+from .metadata_value import MetadataValue
+
+if TYPE_CHECKING:
+    from dagster._core.definitions.assets import AssetsDefinition, SourceAsset
+    from dagster._core.definitions.cacheable_assets import CacheableAssetsDefinition
+
+DEFAULT_SOURCE_FILE_KEY = "asset_definition"
+
+
+@experimental
+@whitelist_for_serdes
+class LocalFileCodeReference(DagsterModel):
+    """Represents a local file source location."""
+
+    file_path: str
+    line_number: int
+    label: Optional[str] = None
+
+
+@experimental
+@whitelist_for_serdes
+class UrlCodeReference(DagsterModel):
+    """Represents a source location which points at a URL, for example
+    in source control.
+    """
+
+    url: str
+    label: Optional[str] = None
+
+
+@experimental
+@whitelist_for_serdes
+class CodeReferencesMetadataValue(DagsterModel, MetadataValue["CodeReferencesMetadataValue"]):
```

Comment:
I think this is showing as new because the stacked PRs were rebased. See [this PR](https://github.com/dagster-io/dagster/pull/21465#discussion_r1587832484) for brief naming discussion - personally I bias towards `References` because it makes clear that the value can hold more than one reference. When I see `reference` in the wild I typically interpret it as singular unless it's used as an adjective (e.g. `reference material`).